woocommerce/woocommerce
# Add JS remote logging package

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes #49284.


This PR introduces a new JavaScript package for remote error logging in WooCommerce. The package captures and logs JS errors, sending them to our centralized logging system. 

Key features include:

- Automatic capture of `window.onerror` and `unhandledrejection` events
- A manual log function for unusual events
- Error filtering to exclude third-party stack frames
- Rate limiting to prevent API flooding
- Customizable through WordPress filters

See [REDAME.md ](https://github.com/woocommerce/woocommerce/blob/add/js-remote-error-logging/packages/js/remote-logging/REDAME.md) for more details.

This PR only adds a new package, it does not change any existing functionality. We will have a separate PR to integrate the new package with WC Admin or WC Blocks to capture and log errors from the WC JavaScript code.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**For QA team, this PR only adds a new package, it does not change any existing functionality so you can skip doing manual testing for this PR.** 
 
1. Sandbox public-api.wordpress.com and use D156767-code patch if it has not been merged yet.
2. Checkout out [https://github.com/woocommerce/woocommerce/tree/test/wcadmin-remote-js-logging](wcadmin-remote-js-logging) branch or cherry-pick https://github.com/woocommerce/woocommerce/commit/093ba079fd84d90fd4c301a39db11e86c225a17e
3. Run `pnpm install`
4. Run `pnpm watch:build` on the `./plugins/woocommerce` folder.
5. Use local fresh site or delete `woocommerce_allow_tracking` option to test the changes.
6. Open browser dev console and run `window.localStorage.setItem( 'debug', 'wc:remote-logging')` to enable the debug logging.
7. Go to `WooCommerce > Home`
5. Confirm `wc:remote-logging Tracks is not enabled.` message is logged in the console.
6. Enable Tracks by adding `wp option set woocommerce_allow_tracking yes`
8. Reload the page
9. Confirm `wc:remote-logging Caught error event` and `wc:remote-logging Sending error to API` messages are logged in the console.
10. Check the network tab in dev console and confirm a `POST` request is made to the remote logging API.
11. Reload the page
12. Confirm `Rate limit reached. Skipping send error` message is logged in the console.
13. Comment out `setTimeout` in [`plugins/woocommerce-admin/client/lib/init-remote-logging.ts`](https://github.com/woocommerce/woocommerce/blob/093ba079fd84d90fd4c301a39db11e86c225a17e/plugins/woocommerce-admin/client/lib/init-remote-logging.ts) and uncomment `new Promise to test the asynchronous error handling.
14. Reload the page
15. Confirm `wc:remote-logging Caught unhandled rejection:` message is logged in the console.
16. Check the network tab in the dev console and confirm a `POST` request is made to the remote logging API.
17. Comment out `new Promis` in `plugins/woocommerce-admin/client/lib/init-remote-logging.ts`
18. Uncomment `log` in `plugins/woocommerce-admin/client/lib/init-remote-logging.ts`
19. Reload the page
20. Confirm `wc:remote-logging Sending log to API` message is logged in the console.
21. Check the network tab in dev console and confirm a `POST` request is made to the remote logging API.
22.  Please check `D156767` patch for how to view the logs in Kibana.

<img src="https://github.com/user-attachments/assets/aa5740b9-d189-4194-85d6-8685dea7ec91" width="600px" />

Error log:
![Screenshot 2024-07-29 at 17 54 02](https://github.com/user-attachments/assets/89420a7b-88e7-43d2-a19a-bd70e18f5205)

Info log:
![Screenshot 2024-07-29 at 17 54 59](https://github.com/user-attachments/assets/f794b1bb-65ed-4111-a3ee-31dda13cbd73)





### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

Add JS remote logging package

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
